doc/operations/code_upgrade.rst

```diff
@@ -0,0 +1,189 @@
+
+4. The original nodes (``Node 0``, ``Node 1`` and ``Node 2``) can then safely be retired.
+
+- ``Node 0`` is retired, 5 nodes remaining, ``f = 2``:
+
+.. mermaid::
+
+    graph TB;
+        classDef NewNode fill:Turquoise
+        classDef RetiredNode fill:LightGray
+        classDef Primary stroke-width:4px
+
+        Node0((Node 0))
+        class Node0 RetiredNode
+
+        subgraph Service
+            subgraph Old Nodes
+                Node1((Node 1))
+                class Node1 Primary
+                Node2((Node 2))
+            end
+
+            subgraph New Nodes
+                Node3((Node 3))
+                Node4((Node 4))
+                Node5((Node 5))
+                class Node3 NewNode
+                class Node4 NewNode
+                class Node5 NewNode
+            end
+        end
+
+- ``Node 1`` (primary) is retired, 4 nodes remaining, ``f = 1``. ``Node 4`` becomes primary after election phase (during which service cannot temporarily process requests that mutate the state of the key-value store):
+
+.. mermaid::
+
+    graph TB;
+        classDef NewNode fill:Turquoise
+        classDef RetiredNode fill:LightGray
+        classDef Primary stroke-width:4px
+
+        Node0((Node 0))
+        Node1((Node 1))
+        class Node0 RetiredNode
+        class Node1 RetiredNode
+
+        subgraph Service
+            subgraph Old Nodes
+                Node2((Node 2))
+            end
+
+            subgraph New Nodes
+                Node3((Node 3))
+                Node4((Node 4))
+                class Node4 Primary
+                Node5((Node 5))
+                class Node3 NewNode
+                class Node4 NewNode
+                class Node5 NewNode
+            end
+        end
+
+.. note:: It is possible for another old node (e.g. ``Node 2``) to become primary when the old primary node is retired. However, eventually, the primary-ship of the service will be transferred to one of the new nodes (e.g. ``Node 4``):
+
+- ``Node 2`` is retired, 3 nodes remaining, ``f = 1``:
+
+.. mermaid::
+
+    graph TB;
+        classDef NewNode fill:Turquoise
+        classDef RetiredNode fill:LightGray
+        classDef Primary stroke-width:4px
+
+        Node0((Node 0))
+        Node1((Node 1))
+        Node2((Node 2))
+        class Node0 RetiredNode
+        class Node1 RetiredNode
+        class Node2 RetiredNode
+
+
+        subgraph Service
+            subgraph New Nodes
+                Node3((Node 3))
+                Node4((Node 4))
+                class Node4 Primary
+                Node5((Node 5))
+                class Node3 NewNode
+                class Node4 NewNode
+                class Node5 NewNode
+            end
+        end
+
+5. Once all old nodes ``0``, ``1`` and ``2`` have been retired, operators can safely stop them:
```

Comment:
It's necessary to learn from the service that retirement has been committed to take that step, the meaning of "retired" needs clarification here.